src/yunohost/permission.py

```diff
@@ -35,309 +36,247 @@
+
+    # Trigger app callbacks
+
+    app = permission.split(".")[0]
 
-        for g in add_group:
-            if 'members' in group_list[g]:
-                allowed_users.union(group_list[g]['members'])
-        for g in del_group:
-            if 'members' in group_list[g]:
-                disallowed_users.union(group_list[g]['members'])
+    old_allowed_users = set(existing_permission["corresponding_users"])
+    new_allowed_users = set(new_permission["corresponding_users"])
 
-        allowed_users = ','.join(allowed_users)
-        disallowed_users = ','.join(disallowed_users)
-        if add_group:
-            hook_callback('post_app_addaccess', args=[app, allowed_users])
-        if del_group:
-            hook_callback('post_app_removeaccess', args=[app, disallowed_users])
+    effectively_added_users = new_allowed_users - old_allowed_users
+    effectively_removed_users = old_allowed_users - new_allowed_users
 
-    return user_permission_list(app, permission)
+    if effectively_added_users:
+        hook_callback('post_app_addaccess', args=[app, ','.join(effectively_added_users)])
+    if effectively_removed_users:
+        hook_callback('post_app_removeaccess', args=[app, ','.join(effectively_removed_users)])
 
+    return new_permission
 
-def user_permission_clear(operation_logger, app=[], permission=None, sync_perm=True):
+
+@is_unit_operation()
+def user_permission_reset(operation_logger, permission, sync_perm=True):
     """
-    Reset the permission for a specific application
+    Reset a given permission to just 'all_users'
 
     Keyword argument:
-        app        -- an application OR sftp, xmpp (metronome), mail
-        permission -- name of the permission ("main" by default)
-        username   -- Username to get informations (all by default)
-        group      -- Groupname to get informations (all by default)
-
+        permission -- Name of the permission (e.g. mail or nextcloud or wordpress.editors)
     """
     from yunohost.hook import hook_callback
     from yunohost.utils.ldap import _get_ldap_interface
     ldap = _get_ldap_interface()
 
-    if permission:
-        if not isinstance(permission, list):
-            permission = [permission]
-    else:
-        permission = ["main"]
+    # By default, manipulate main permission
+    if "." not in permission:
+        permission = permission + ".main"
+
+    # Fetch existing permission
+
+    existing_permission = user_permission_list(full=True)["permissions"].get(permission, None)
+    if existing_permission is None:
+        raise YunohostError('permission_not_found', permission=permission)
+
+    # Update permission with default (all_users)
+
+    operation_logger.related_to.append(('app', permission.split(".")[0]))
+    operation_logger.start()
 
     default_permission = {'groupPermission': ['cn=all_users,ou=groups,dc=yunohost,dc=org']}
+    try:
+        ldap.update('cn=%s,ou=permission' % permission, default_permission)
```

Comment:
We probably should check that we change something in the permission. Because if before to call this function the only group which is allowed is already the group `all_users` LDAP will crash.

Reply:
Indeed, we can add a test to validate the idempotence of the function